### DISCUSSION TITLE
Changing role permissions

### DISCUSSION THREAD

**Discussion by DannyBen** [shardlab/discordrb#115] [Q&A] [answered]
Hello - hoping someone else is here,

I have browsed through the wiki and through the code and its docs - but still unsure what is the best way to do what I want.

I have created a simple `CommandBot` and all I want it to do is to add/remove one or two permissions from a role. In particular, I am trying to create a '!lockdown` command where it just removes the ability to send messages from my `Member` role.

I found `CommandBot#set_role_permission` but unsure it this is the correct place, and if so, what are the required arguments.

Ideally, I was hoping for some syntax close to `bot.roles['Member'].permissions[:send_message] = false`

**Comment by swarley** [comment]
Unfortunately the syntax you're describing isn't really how discord works. To achieve what you want you would likely need to remove all of a user's roles assuming your everyone role doesn't have send messages, or you would need to ensure they don't have any overrides with explicit send messages and apply a role that has an explicit deny send messages override on all channels. Alternatively I recently merged timeout support which is the path given by discord to handle things natively.

**Reply by DannyBen** [reply]
Thank you - well I wish to avoid removing roles, because then a) I will have to re-add them and b) users can obtain this role again.

Timeout might be a good option - I will look for it in the code, but still - I am sure there is *some* way to change the permissions of a role (even if not in my desired DSL-ish syntax) - no?

**Reply by DannyBen** [reply]
Well - I see Timeout is on a per-member basis - if so, I don't see how this helps with lockdown. I wish to not iterate over members - just change a role's permissions.

https://github.com/shardlab/discordrb/blob/a1083b0b944700f582356fec3babd3fe6804c2ca/lib/discordrb/data/member.rb#L135-L139

**Reply by swarley** [reply]
Apologies, I confused myself reading the member as a member and not a role. Yes, there is a way to set role permissions with a DSL. I'll get a snippet to you shortly

**Reply by DannyBen** [reply]
Even without a DSL - I just hoped it is not something that I need to get role ID first, and then calculate bitwise operations.

**Reply by DannyBen** [reply]
I found out how to do it - not sure it is the shortest way, but it seems to be working ok.

```ruby
bot.command(:lock) do |event|
  break unless event.user.id == user_id
  member = event.server.roles.find { |a| a.name == 'Member' }
  member.permissions.can_send_messages = false
end
```

**Reply by swarley** [reply]
Really sorry about that, got caught up in some errands. I would say that's the most efficient way, except that I would recommend using an ID lookup for your roles whenever possible given that names can change. i.e.
```ruby
event.server.role(553953362522144799).permissions.can_send_messages = false
```
If for some reason ID lookup is unfeasible for your use case then the code you have is fine

**Reply by DannyBen** [reply]
Thank you, I will use the ID.

**Comment by swarley** [comment, marked as answer]
For answer purposes see https://github.com/shardlab/discordrb/discussions/115#discussioncomment-1932200
> I found out how to do it - not sure it is the shortest way, but it seems to be working ok.
>
> ```ruby
> bot.command(:lock) do |event|
>   break unless event.user.id == user_id
>   member = event.server.roles.find { |a| a.name == 'Member' }
>   member.permissions.can_send_messages = false
> end
> ```
